test/clj/rems/workflow/test_dynamic.clj

```diff
@@ -191,532 +153,693 @@
 
 (deftest test-add-member
   (let [application (apply-events nil
-                                  [{:event/type :application.event/created
-                                    :event/actor "applicant"
-                                    :application/id 123
-                                    :workflow/type :workflow/dynamic
-                                    :workflow.dynamic/handlers #{"assistant"}}
+                                  [dummy-created-event
                                    {:event/type :application.event/submitted
-                                    :event/actor "applicant"
+                                    :event/time test-time
+                                    :event/actor applicant-user-id
                                     :application/id 123}
                                    {:event/type :application.event/member-added
-                                    :event/actor "applicant"
+                                    :event/time test-time
+                                    :event/actor applicant-user-id
                                     :application/id 123
                                     :application/member {:userid "somebody"}}])
-        injections {:valid-user? #{"member1" "member2" "somebody" "applicant"}}]
-    (testing "add two members"
-      (is (= [{:userid "applicant"} {:userid "somebody"} {:userid "member1"}]
-             (:members
-              (apply-commands application
-                              [{:type :application.command/add-member :actor "assistant" :member {:userid "member1"}}]
-                              injections)))))
+        injections {:valid-user? #{"member1" "member2" "somebody" applicant-user-id}}]
+    (testing "handler can add members"
+      (is (= [{:event/type :application.event/member-added
+               :event/time test-time
+               :event/actor handler-user-id
+               :application/id 123
+               :application/member {:userid "member1"}}]
+             (ok-command application
+                         {:type :application.command/add-member
+                          :actor handler-user-id
+                          :member {:userid "member1"}}
+                         injections))))
     (testing "only handler can add members"
       (is (= {:errors [{:type :forbidden}]}
-             (handle-command {:application-id 123 :time test-time
-                              :type :application.command/add-member :actor "applicant" :member {:userid "member1"}}
-                             application
-                             injections)
-             (handle-command {:application-id 123 :time test-time
-                              :type :application.command/add-member :actor "member1" :member {:userid "member2"}}
-                             application
-                             injections))))
+             (fail-command application
+                           {:type :application.command/add-member
```

Comment:
hmm this is not a valid command without :time and :application-id, right?

Reply:
The `fail-command` and `ok-command` helpers add them automatically. `rems.workflow.dynamic/handle-command` also validates all commands, so it would detect such issues.

Reply:
right!